marlowe-playground-client/src/Marlowe/Blockly.purs

```diff
@@ -999,278 +1001,354 @@ toDefinition blockType@(ValueType UseValueValueType) =
-  mkBlock t = block [ blockType (show t) ] []
+--
+-- Code generation
+--
+blockToContract :: BDom.Block -> Either String (Term Contract)
+blockToContract block = lmap explainParseTermError $ blockToTerm block
 
-workspaceBlocks :: forall a b. HTML a b
-workspaceBlocks =
-  xml [ id_ "workspaceBlocks", style "display:none" ]
-    [ block [ blockType (show BaseContractType), x "13", y "187", id_ rootBlockName ] []
-    ]
+class BlockToTerm a where
+  blockToTerm :: forall m. MonadError ParseTermError m => BDom.Block -> m (Term a)
 
-parse :: forall a. Parser a -> String -> Either String a
-parse p = lmap show <<< runParser' (parens p <|> p)
+data ParseTermError
+  = InvalidBlock BDom.Block String
+  | MissingArgument BDom.Block String
+  | InvalidChildType String
+  | InvalidFieldCast String String
+  | ErrorInChild BDom.Block String ParseTermError
 
-buildGenerator :: Blockly -> Effect Generator
-buildGenerator blockly = do
-  generator <- mkGenerator blockly "Marlowe"
-  let
-    mkGenFun :: forall a t. Show a => Show t => t -> (Block -> Either String a) -> Effect Unit
-    mkGenFun blockType f = insertGeneratorFunction generator (show blockType) (rmap show <<< f)
-  traverse_ (\t -> mkGenFun t (baseContractDefinition generator)) [ BaseContractType ]
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) contractTypes
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) payeeTypes
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) partyTypes
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) tokenTypes
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) observationTypes
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) valueTypes
-  traverse_ (\t -> mkGenFun t (blockDefinition t generator)) actionTypes
-  traverse_ (\t -> mkGenFun t (boundsDefinition generator)) [ BoundsType ]
-  pure generator
-
-class HasBlockDefinition a b | a -> b where
-  blockDefinition :: a -> Generator -> Block -> Either String b
-
-baseContractDefinition :: Generator -> Block -> Either String (Term Contract)
-baseContractDefinition g block = case statementToCode g block (show BaseContractType) of
-  Either.Left _ -> pure $ Hole "contract" Proxy NoLocation
-  Either.Right s -> parse (mkDefaultTerm <$> Parser.contract) s
-
-getAllBlocks :: Block -> Array Block
-getAllBlocks currentBlock =
-  currentBlock
-    : ( case nextBlock currentBlock of
-          Nothing -> []
-          Just nextBlock' -> getAllBlocks nextBlock'
-      )
+explainParseTermError :: ParseTermError -> String
+explainParseTermError (InvalidBlock (BDom.Block block) expectedType) = "Block with id " <> show block.id <> " and type " <> show block.type <> " is not a valid " <> expectedType <> " type"
+
+explainParseTermError (MissingArgument (BDom.Block block) attr) = "Block " <> show block.id <> " is missing required argument " <> show attr
+
+explainParseTermError (InvalidChildType childType) = "Attribute should be of type " <> childType
```

Comment:
why not just create a `Show ParseTermError` instance?

Reply:
I don't like to overuse Show semantics... they are normally used for generally displaying values in a "sensible" way, so you end up with things like the following in the reachability results.

```
((WhenCasePath 0) : Nil)
```
In personal projects I've defined an Explain class that has the same signature which I use to explain errors in a human readable way. I'm happy to define that here.

Slightly related, I'd like to try these in our tests to avoid having to define Show instances for types that we don't really want to show.

https://github.com/hdgarrood/purescript-debugged

Bonus, it has a blog series called "Down with Show" :P

Reply:
sounds good to me